room/compiler-processing/src/main/java/androidx/room/compiler/processing/ksp/KSTypeExt.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2020 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.room.compiler.processing.ksp
+
+import com.squareup.javapoet.ClassName
+import com.squareup.javapoet.ParameterizedTypeName
+import com.squareup.javapoet.TypeName
+import org.jetbrains.kotlin.ksp.symbol.KSClassifierReference
+import org.jetbrains.kotlin.ksp.symbol.KSDeclaration
+import org.jetbrains.kotlin.ksp.symbol.KSName
+import org.jetbrains.kotlin.ksp.symbol.KSType
+import org.jetbrains.kotlin.ksp.symbol.KSTypeReference
+
+// catch-all type name when we cannot resolve to anything.
+private val UNDEFINED = ClassName.get("androidx.room.compiler.processing.kotlin.error", "Undefined")
+
+/**
+ * Turns a KSTypeReference into a TypeName
+ *
+ * We try to achieve this by first resolving it and iterating.
+ * If some types cannot be resolved, we do a best effort name guess from the KSTypeReference's
+ * element.
+ */
+internal fun KSTypeReference?.typeName(): TypeName {
+    if (this == null) {
+        return UNDEFINED
+    }
+    val resolved = resolve()
+    return resolved?.typeName() ?: fallbackClassName()
+}
+
+private fun KSTypeReference.fallbackClassName(): ClassName {
+    return (element as? KSClassifierReference)?.let {
+        ClassName.get("", it.referencedName())
+    } ?: UNDEFINED
+}
+
+private fun KSName.typeName(): ClassName? {
+    if (asString().isBlank()) {
+        // fallback to reference
+        return null
+    }
+    return ClassName.get(getQualifier(), getShortName())
```

Comment:
How does this work for nested classes? For `package.Foo.Bar`what does `getQualifier()` return?

Reply:
Currently KSName.getQualifier returns everything before the last "."

Reply:
Then it sounds like this won't properly handle nested classes, since for `my.package.Foo.Bar` you want to do `ClassName.get("my.package", "Foo", "Bar")` and `getQualifier` is not quite right.

Reply:
On a side note it sounds like an implementation detail of Java/Kotlin Poet. We do have a tracking issue and plan to address it soon https://github.com/android/kotlin/issues/23

Reply:
i do have a test for that:
https://github.com/yigit/androidx/pull/2/files#diff-f2dd95dfd9642a1a5b721fcb0382375fR43

but of course, the equals in javapoet might be just failing to discover the difference. I'll investigate.

Reply:
i'm guessing i'll need proper support from KSP before fixing this.

Reply:
Some update: After looking into the kotlin spec regarding names, dot based name is standard at language perspective and slash based names actually is a special case for JVM platform runtime name. We've added an API  `KSDeclaration.packageName` to work with your use case, but you will need to make some adjustments on the use site, like in this example, it's best to have a new function working on a declaration rather than a ClassName.

Reply:
updated using the new API.
it is still a bit weird implementation where i need to get the package name and then subtract it from the qualified name.
but does what we need, thanks!